File: Algorithms/mode.lua
--------------------------------------------------
-- Math Algorithms                              --
-- https://github.com/JoseCintra/MathAlgorithms --
--------------------------------------------------

-- Name: Statistical Mode
-- Purpose: Calculates the statistical mode of a data set using maps as a data structure
-- References: https://en.wikipedia.org/wiki/Mode_(statistics)
-- Language: Lua
-- Year: 2024
-- Author: José Cintra 
-- Run Online at: 

io.write("\nStatistical Mode\n")
io.write("------------------\n")
data = {2,3,9,7,5,3,7,3}
mode = {}
l = table.getn(data)
for i = 1, l, 1 do
  if mode[data[i]] == nil then
    mode[data[i]] = 1
  else
    mode[data[i]] = mode[data[i]] + 1
  end
end
for k,v in next,mode do
  io.write(k,v, "\n")
end  

File: Algorithms/Mode.dart
void main() {
  print('Statistical Mode');
  print('----------------');
  var data = [2, 3, 9, 7, 5, 3, 7, 3];
  var mode = new Map();
  data.forEach((e) {
    if (!mode.containsKey(e)) {
      mode[e] = 1;
    } else {
      mode[e] = mode[e] + 1;
    }
  });
  var sortedMode = Map.fromEntries(
      mode.entries.toList()..sort((e1, e2) => e2.value - e1.value));
  print(sortedMode);
}
